### PR TITLE
Scaffold location for open feature tests.

### PR DESCRIPTION
**Pull request by kinyoklion**
Drafting this for some questions.

I want to determine if this approach is favorable.

There are broadly a few ways to implement open feature tests.
- Implement them as a sdkKind similar to client/server. I think this makes the most sense from an external perspective. It is a different kind of thing. It is based on a server SDK though. What this PR does.
- Implement a capability that would exercise the open feature methods and could be ran in combination with other server tests.

Independent of this there are a couple ways to handle evaluation parameters/results. 
- Implement specific types for the parameters and result. The method I have taken so far. There is some duplicate work.
- Implement a translation that provides the results in a way compatible with the existing evaluation. Benefit being we could run single evaluations (not all flags) that are specified in existing suites. The big down side is most of OpenFeature work is converting these types, so it makes it hard to know if that is working completely correctly.

### CODE REVIEW

File: mockld/streaming_service.go [hunk 1]
```diff
@@ -78,6 +78,8 @@ func NewStreamingService(
 	streamHandler := streams.Handler(allDataChannel)
 	router := mux.NewRouter()
 	switch sdkKind {
+	case ServerSideOpenFeature:
+		fallthrough
```
Comment (kinyoklion) [inline mockld/streaming_service.go:82]: Could copy/paste if that is more favorable.

File: sdktests/server_side_open_feature_eval.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+	"github.com/launchdarkly/sdk-test-harness/framework/ldtest"
+	"github.com/launchdarkly/sdk-test-harness/mockld"
+	"github.com/launchdarkly/sdk-test-harness/servicedef"
+	"gopkg.in/launchdarkly/go-sdk-common.v2/ldvalue"
+	"gopkg.in/launchdarkly/go-server-sdk-evaluation.v1/ldbuilders"
+)
+
+func doServerSideOpenFeatureEvalTests(t *ldtest.T) {
+	expectedValueV1 := ldvalue.Int(1)
```
Comment (kinyoklion) [inline sdktests/server_side_open_feature_eval.go:13]: I've only put a test directly here to verify that things happen. This is not an example of an actual setup. I would make the common files, but I wanted to validate the other elements before doing that.

File: sdktests/testapi_sdk_client.go [hunk 3]
```diff
@@ -209,6 +209,22 @@ func (c *SDKClient) EvaluateFlag(t *ldtest.T, params servicedef.EvaluateFlagPara
 	return resp
 }
 
+func (c *SDKClient) EvaluateOpenFeatureFlag(t *ldtest.T, params servicedef.OpenFeatureEvaluateFlagParams) servicedef.OpenFeatureEvaluateFlagResponse {
```
Comment (kinyoklion) [inline sdktests/testapi_sdk_client.go:212]: Maybe I could make this less duplicated with generics?

**Comment by kinyoklion** [comment]
I just want to get feedback on the approach at the moment. Not a specific test.

**Comment by eli-darkly** [comment]
I have mixed feelings about the basic premise of this— that is, reusing `sdk-test-harness` for OpenFeature testing.

I understand the desire to avoid reinventing the wheel, and the basic concept of contract testing seems like it could be applicable to these libraries.

However... what we would be testing has a different shape. It's a different abstraction, even if it's an LD SDK under the covers. The entire sdk-test-harness protocol is designed for the SDK model— so, for instance, the evaluate command takes a _user_. Would the test service for an OpenFeature library take a user, and then transform that into an OF evaluation context, and pass that to the OF client which would then end up transforming it back into a user? Or would we be adding new commands to the protocol which is _only_ relevant for OpenFeature clients, and saying that if it's an OpenFeature client then we will never send any of the other commands that are for the SDKs, and the test suites will be implemented totally separately?

If so, I'm not sure how much value we would be getting from implementing these similar but incompatible things in the same project. If on the other hand we just wanted to reuse the lower-level concepts, like the basic HTTP mechanism and the `ldtest` framework... those are easily reusable by another project. That's why they have their own parent package, `framework`. My goal with that was to eventually get rid of the equivalent stuff in `sse-test-harness`, and have `sse-test-harness` simply import those packages from `sdk-test-harness`.

**Comment by kinyoklion** [comment]
> I have mixed feelings about the basic premise of this— that is, reusing `sdk-test-harness` for OpenFeature testing.
> 
> I understand the desire to avoid reinventing the wheel, and the basic concept of contract testing seems like it could be applicable to these libraries.
> 
> However... what we would be testing has a different shape. It's a different abstraction, even if it's an LD SDK under the covers. The entire sdk-test-harness protocol is designed for the SDK model— so, for instance, the evaluate command takes a _user_. Would the test service for an OpenFeature library take a user, and then transform that into an OF evaluation context, and pass that to the OF client which would then end up transforming it back into a user? Or would we be adding new commands to the protocol which is _only_ relevant for OpenFeature clients, and saying that if it's an OpenFeature client then we will never send any of the other commands that are for the SDKs, and the test suites will be implemented totally separately? In your example code here, it looks like you're doing the latter.
> 
> If so, I'm not sure how much value we would be getting from implementing these similar but incompatible things in the same project. If on the other hand we just wanted to reuse the lower-level concepts, like the basic HTTP mechanism and the `ldtest` framework... those are easily reusable by another project without copying anything. That's why they have their own parent package, `framework`. My goal with that was to eventually get rid of the equivalent stuff in `sse-test-harness`, and have `sse-test-harness` simply import those packages from `sdk-test-harness`.

So

> I have mixed feelings about the basic premise of this— that is, reusing `sdk-test-harness` for OpenFeature testing.
> 
> I understand the desire to avoid reinventing the wheel, and the basic concept of contract testing seems like it could be applicable to these libraries.
> 
> However... what we would be testing has a different shape. It's a different abstraction, even if it's an LD SDK under the covers. The entire sdk-test-harness protocol is designed for the SDK model— so, for instance, the evaluate command takes a _user_. Would the test service for an OpenFeature library take a user, and then transform that into an OF evaluation context, and pass that to the OF client which would then end up transforming it back into a user? Or would we be adding new commands to the protocol which is _only_ relevant for OpenFeature clients, and saying that if it's an OpenFeature client then we will never send any of the other commands that are for the SDKs, and the test suites will be implemented totally separately? In your example code here, it looks like you're doing the latter.
> 
> If so, I'm not sure how much value we would be getting from implementing these similar but incompatible things in the same project. If on the other hand we just wanted to reuse the lower-level concepts, like the basic HTTP mechanism and the `ldtest` framework... those are easily reusable by another project without copying anything. That's why they have their own parent package, `framework`. My goal with that was to eventually get rid of the equivalent stuff in `sse-test-harness`, and have `sse-test-harness` simply import those packages from `sdk-test-harness`.

I think the biggest thing that is re-used is that all the configuration for the SDK is the same for OpenFeature as it is for an LD SDK. Because we are directly creating the SDK. But that could be shared in some way as well. It isn't a test for any open feature provider implementation, only for LD SDKs.

As the only capability of an open feature client itself is evaluation, I was only focusing on just evaluation. In which case the test takes an evaluation context, not a user, because that is the shape of things for OpenFeature. In the provider though that gets converted to a user. and that could be verified in events.

That said, if you do an evaluation against a context, then the events also should come out the same.

**Pull request closed without merging**